fix includes for framework libs

jacquetc opened the pull request:
Hello,

This is a tentative fix for using Qt's "framework" libs on MacOS. See #450

ahayzen-kdab commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -319,6 +343,10 @@ impl QtBuild {
 
         match std::fs::read_to_string(prl_path) {
             Ok(prl) => {
+                println!("cargo:rustc-link-lib=reading {}", prl_path);
```
i think you shouldn't have things like `reading` here (why CI is exploding) ? Or are you trying to debug what is happening at each step ?

jacquetc replied:
Yes, it was exactly that. I could push a commit cleaned of these prints.

jacquetc replied:
removed in last commit

ahayzen-kdab commented on .gitignore:
```diff
@@ -20,3 +20,5 @@ compile_commands.json
 .cmps/
 
+# macOS
+.DS_Store
```
note you could put this in your global gitignore at `~/.gitignore` or `~/.config/git/ignore` (depending on config) instead of per project

jacquetc replied:
Yet, some .gitignore lines are so common. Without that, there is a risk of such system artifacts being commited by inadvertence. There are temp files like those finishing by ~

jacquetc replied:
removed in last commit. If someone decide to have a more complete .gitignore, there are premade models on github

jacquetc commented:
> 

I was trying to fix primarily the use of Qt from the official installer. So, the use of framework libs, so adding a few specific args to the compiler. 

So, my method was : 
- detect if the targeted Qt use framework libs
- if so, adapt the args.

I'll be able to recreate the initial problem at home tonight. Yet, it was the same error than in #450

ahayzen-kdab commented:
> I was trying to fix primarily the use of Qt 6.5 from the official installer. So, the use of framework libs, so adding a few specific args to the compiler.

Awesome thanks ! There are just so many different combinations it's useful to know which is being investigated :-) I find it fun that we solved use of frameworks from a homebrew install but not from a the Qt official installer, so it will be interesting to see the difference :-) Any help on macOS or Windows builds is much appreciated as we are mostly Linux devs :sweat_smile:

jacquetc commented:
> > I was trying to fix primarily the use of Qt 6.5 from the official installer. So, the use of framework libs, so adding a few specific args to the compiler.
> 
> Awesome thanks ! There are just so many different combinations it's useful to know which is being investigated :-) I find it fun that we solved use of frameworks from a homebrew install but not from a the Qt official installer, so it will be interesting to see the difference :-) Any help on macOS or Windows builds is much appreciated as we are mostly Linux devs 😅

I'm a cross-platform Qt dev and DevOps, so I know a little bit about such issues. I'll try to help, but must be clear about the platforms and installs I can play with:
- Apple M1 - homebrew
- Apple M1 - official installer
- Apple M1 - vcpkg (a cxx-qt project consuming qt from vcpkg)
- Windows ARM - official installer
- Windows ARM - vcpkg (a cxx-qt project consuming qt from vcpkg)
- Linux Ubuntu / Fedora ARM - system Qt
- Linux Ubuntu / Fedora ARM - official installer
- Linux Ubuntu / Fedora ARM - vcpkg  (a cxx-qt project consuming qt from vcpkg)

I found out reading the code, that my first problem was to make qmake visible. So adding its folder to the PATH is a prerequise before using cxx-qt.

ahayzen-kdab commented:
> I'm a cross-platform Qt dev and DevOps, so I know a little bit about such issues.

Awesome :-)

> ... make qmake visible. So adding its folder to the PATH is a prerequise before using cxx-qt.

Right, we also have a `QMAKE` env var you can set to point to a custom `qmake`, either `qmake` needs to be in your path or you set `QMAKE` to point to it.

jacquetc commented:
I would use another name for the variable like `CXX_QT_QMAKE`. Adding a prefixes is to not interfere with QT's future possible env variable. It's your choice of course.

jacquetc commented:
### State of local tests:
_Do not merge before the completion of this list !_

- [x] on Apple M1 with /homebrew
- [x] on Apple M1 with official Qt installer
- [ ] Apple M1 - vcpkg (a cxx-qt project consuming qt from vcpkg)
- [ ] Windows ARM - official installer
- [ ] Windows ARM - vcpkg (a cxx-qt project consuming qt from vcpkg)
- [ ] Linux Ubuntu / Fedora ARM - system Qt
- [ ] Linux Ubuntu / Fedora ARM - official installer
- [ ] Linux Ubuntu / Fedora ARM - vcpkg (a cxx-qt project consuming qt from vcpkg)

jacquetc commented:
> Also note how previously we battled adding support for frameworks from homebrew installs here [c6ac928](https://github.com/KDAB/cxx-qt/commit/c6ac92851e7278f518c3c3ea17326851e6ff854f)
> 
> Which route of Qt install are you trying to improve here ? As i thought the remaining issues were around using the Qt official binaries ? Or have you found another scenario of frameworks that doesn't work ? 😅
> 
> Maybe it would be useful if you could post the error that occurs on your machine with `main` so that we know which of the various combinations you are improving here :-)

I compared the Qt installation from homebrew and Qt official installation.
I can be mistaken, but it seems that the work done in [c6ac928](https://github.com/KDAB/cxx-qt/commit/c6ac92851e7278f518c3c3ea17326851e6ff854f) had an unintentional side effect to make it compile with homebrew's installation, but by adding support for the framework style libs. 

Homebrew installs Qt quite traditionally, plus add links like /opt/homebrew/include/QtCore linking to /opt/Cellar/qt/6.5.1_2/include/QtCore . So, somewhere, certainly [this line](https://github.com/KDAB/cxx-qt/blob/c6ac92851e7278f518c3c3ea17326851e6ff854f/crates/qt-build-utils/src/lib.rs#L311C1-L311C7), the good "include" path from homebrew was added. Yet, the work done was a necessary step which I was able to build on. 

In the Qt official installation, "packages" containing the libraries and the headers are found in [...]/Qt/6.5.2/macos/lib/[module name].framework
Ex: 
[...]/Qt/6.5.2/macos/lib/QtCore.framework
_contains the files:_
QtCore (links to Versions/A/QtCore)
Headers (links to Versions/A/Headers)
Resources (links to Versions/A/Resources)
Versions/A/QtCore
Versions/A/Headers
Versions/A/Resources

I added the include paths like [...]/Qt/6.5.2/macos/lib/QtCore.framework/Headers
And I added the framework args -f  , like -f[...]/Qt/6.5.2/macos/lib/QtCore.framework

ahayzen-kdab commented:
> I can be mistaken, but it seems that the work done in [c6ac928](https://github.com/KDAB/cxx-qt/commit/c6ac92851e7278f518c3c3ea17326851e6ff854f) had an unintentional side effect to make it compile with homebrew's installation, but by adding support for the framework style libs.

Yes this change was to add support for framework style libs and specifically what homebrew was producing, but were not yet checked for what the Qt installer produces (which appears to be subtly different as you've stated).

If you get a chance, checking that Qt 5.15 works via the Qt installer on macOS could would be good in case that is different again.

ahayzen-kdab commented on .gitignore:
```diff
@@ -18,5 +18,4 @@ rustup/*
 # cmps (https://github.com/leonmatthes/cmps)
-.cmps/
-
+.cmps/
```
likely a new line missing at the end of the file here which is causing a diff

jacquetc replied:
```suggestion
.cmps/

```

ahayzen-kdab commented on crates/cxx-qt-lib/build.rs:
```diff
@@ -187,6 +188,10 @@ fn main() {
         cxx_build::CFG
             .exported_header_dirs
             .push(include_path.as_path());
+        println!(
```
this looks like a trace line ?

jacquetc replied:
I used that to debug includes, of course. But isn't useful information ? Maybe in some other form ?

jacquetc replied:
fixed in the next comment

ahayzen-kdab commented on crates/cxx-qt-lib/build.rs:
```diff
@@ -271,5 +276,6 @@ fn main() {
     // GCC + Clang
     builder.flag_if_supported("-std=c++17");
 
+    println!("cargo:rerun-if-env-changed=compile");
```
this looks like a trace line ?

jacquetc replied:
Isn't it useful to have markers at least in the higher level class to follow the steps ?

jacquetc replied:
```suggestion
```

ahayzen-kdab commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -232,6 +234,22 @@ impl QtBuild {
 
+        fn determine_if_has_framework_libs(qmake_executable: &String) -> bool {
+            let lib_path = QtBuild::static_qmake_query(qmake_executable, "QT_INSTALL_LIBS");
+            let target: Result<String, env::VarError> = env::var("TARGET");
```
do we need the type defined here? can't it just be `let target = env::var("TARGET");` and let the compiler determine the type?

jacquetc replied:
Of course, it's inferred. I tend to type variables when in unknown code.
```suggestion
            let target = env::var("TARGET");
```

ahayzen-kdab commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -350,7 +381,7 @@ impl QtBuild {
     ) -> String {
         for arch in ["", "_arm64-v8a", "_armeabi-v7a", "_x86", "_x86_64"] {
-            let prl_path = format!(
+            let prl_path: String = format!(
```
do we need this type? doesn't `format!(...)` always go to String?

jacquetc replied:
Force of habit from C++, typing all, sorry
```suggestion
            let prl_path = format!(
```

ahayzen-kdab commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -369,10 +400,12 @@ impl QtBuild {
         }
 
-        format!(
+        let prl_file: String = format!(
```
same here

jacquetc replied:
Same reason
```suggestion
        let prl_file = format!(
```

ahayzen-kdab commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -381,7 +414,7 @@ impl QtBuild {
         println!("cargo:rustc-link-search={lib_path}");
 
-        let target = env::var("TARGET");
+        let target: Result<String, env::VarError> = env::var("TARGET");
```
likely don't need this change

jacquetc replied:
```suggestion
```

ahayzen-kdab commented on crates/qt-build-utils/src/parse_cflags.rs:
```diff
@@ -196,6 +196,7 @@ pub(crate) fn parse_libs_cflags(name: &str, link_args: &[u8], _builder: &mut cc:
                         } else {
                             match extract_lib_from_filename(target, &file_name) {
                                 Some(lib_basename) => {
+                                    println!("cargo:rustc-link-search=lib_from_filename :",);
```
trace line?

jacquetc replied:
Yes, but I thought that it was useful. I'll remove it
```suggestion
```

ahayzen-kdab commented:
It passes CI now :partying_face: 

I'll let @Be-ing comment on the changes to the build side as well as he has been working on this side the most.

jacquetc commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -191,6 +192,7 @@ impl QtBuild {
                 .output()
             {
                 Err(e) if e.kind() == std::io::ErrorKind::NotFound => Err(QtBuildError::QtMissing),
+
```
I'll try to be more careful on these next times.
```suggestion
```

jacquetc commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -417,6 +450,11 @@ impl QtBuild {
             };
 
+            println!(
+                "cargo:rustc-link-search=using {} from {}",
```
```suggestion
```

jacquetc replied:
fixed int the next comment

jacquetc commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -417,6 +450,11 @@ impl QtBuild {
                 )
             };
 
+            println!(
+                "cargo:rustc-link-search=using {} from {}",
+                link_lib, prl_path
+            );
+
```
```suggestion
```

jacquetc commented:
> It passes CI now 🥳
> 
> I'll let @Be-ing comment on the changes to the build side as well as he has been working on this side the most.

Thank you for your help @ahayzen-kdab

jacquetc commented on crates/cxx-qt-lib/build.rs:
```diff
@@ -187,6 +188,10 @@ fn main() {
         cxx_build::CFG
             .exported_header_dirs
             .push(include_path.as_path());
+        println!(
+            "cargo:rerun-if-changed=include_path: {}",
+            include_path.display()
+        );
```
```suggestion
```

jacquetc commented:
It seems that there are problem with the VCPKG version of Qt (told [here](https://github.com/KDAB/cxx-qt/issues/450#issuecomment-1655416342)) with this branch. So please wait for proper testing. 

I have troubles installing VCPKG's Qt on Linux & MacOS due to "ICU" not finding "autoconf-archive". Investigating... Sure, it's outside our scope.

Be-ing commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -448,12 +480,21 @@ impl QtBuild {
+            }
+        }
+        let root_path: String = self.qmake_query("QT_INSTALL_HEADERS");
         for qt_module in &self.qt_modules {
             paths.push(format!("{root_path}/Qt{qt_module}"));
         }
```
Is this correct when Qt is installed as a framework? This allows using include paths like `#include <QString>` instead of `#include <QtCore/QString>`. I would think we'd need to add `{lib_root_path}/Qt{qt_module}.framework/Headers/Qt{qt_module}` in the framework case.

jacquetc replied:
I never included using `#include <QtCore/QString>`, always using `#include <QString>` like it's shown in the Qt documentation. So, I don't see the problem. Or I misunderstand your comment ?

Be-ing replied:
I think this code should likely be moved into an `else` block of the `if self.has_framework_libs` statement. If it wasn't sufficient before to get the include paths with frameworks, it's probably okay to not add these include paths when using frameworks.

jacquetc replied:
I did a few researches :
The module-style includes are essential in MacOS framework style. Using the old #include <QtCore/qobject.h> or #include <QtCore/QObject> is not possible any longer. The old style still runs well on Windows and Linux thought.

I was testing only on qml_minimal example project, so this problem didn't appear.
On a hunch, I ran `QMAKE=/Users/cyril/Qt/6.5.2/macos/bin/qmake cargo run -p qml-minimal-no-cmake` , so Qt's official install on MacOs with framework style

It gave me these errors:
  cargo:warning=In file included from /Users/cyril/Devel/cxx-qt/target/debug/build/cxx-qt-lib-c19809693f328b69/out/cxxbridge/sources/cxx-qt-lib/src/core/qlist/qlist_i32.rs.cc:1:
  cargo:warning=/Users/cyril/Devel/cxx-qt/target/debug/build/cxx-qt-lib-c19809693f328b69/out/cxx-qt-lib/qlist.h:11:10: fatal error: 'QtCore/QList' file not found
  cargo:warning=#include <QtCore/QList>
  cargo:warning=         ^~~~~~~~~~~~~~
  cargo:warning=1 error generated.
  exit status: 1

ahayzen-kdab replied:
Huh curious, i found doing `#include <QtCore/QObject>` useful as it allows you to quickly determine which modules you need to include and link or why a library needs eg QtQuick as you can search for it quickly.

Guess we'll have to drop doing that if this doesn't work in macOS anymore (i can create a separate PR to do this), unless there is a different include folder that can be used that does include the module name + the header name ?

jacquetc replied:
Don't be hasty, the truth can be more complex that my explanation :D I just fixed all the QtCore, QtGui, QtQml in your lib. Guess what ? Qt's itself is floundering :
```
  exit status: 1
  cargo:warning=In file included from /Users/cyril/Devel/cxx-qt/target/debug/build/cxx-qt-lib-c19809693f328b69/out/cxxbridge/sources/cxx-qt-lib/src/core/qlist/qlist_i32.rs.cc:1:
  cargo:warning=In file included from /Users/cyril/Devel/cxx-qt/target/debug/build/cxx-qt-lib-c19809693f328b69/out/cxx-qt-lib/qlist.h:11:
  cargo:warning=In file included from /Users/cyril/Qt/6.5.2/macos/lib/QtCore.framework/Headers/QList:1:
  cargo:warning=/Users/cyril/Qt/6.5.2/macos/lib/QtCore.framework/Headers/qlist.h:8:10: fatal error: 'QtCore/qarraydatapointer.h' file not found
  cargo:warning=#include <QtCore/qarraydatapointer.h>
  cargo:warning=         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.
  exit status: 1
```

So, something is lacking.

ahayzen-kdab replied:
OK, i'll hold off that mega change for now :sweat_smile: And as it looks like Qt itself is doing the same, looks like it should be supported which is good :-D

jacquetc replied:
I'm wondering if there isn't a link with the "-framework [Qt module]" argument. My clang doesn't have it, yet it's a common argument for frameworks on the internet.

Be-ing replied:
> I'm wondering if there isn't a link with the "-framework [Qt module]" argument. My clang doesn't have it, yet it's a common argument for frameworks on the internet.

`parse_libs_cflags` should already take care of linking as a framework if the `.prl` files have those `-framework` or `-F` linker arguments. That's separate from include paths though.

Be-ing replied:
> I did a few researches

Okay, this is helpful information. Let's move
```rust
        let root_path: String = self.qmake_query("QT_INSTALL_HEADERS");
        for qt_module in &self.qt_modules {
            paths.push(format!("{root_path}/Qt{qt_module}"));
        }
        paths.push(root_path);
```
into an `else` block of the `if self.has_framework_libs` statement because that code doesn't work for frameworks.

Regardless, we still have the question of how to support include paths like `#include <QtCore/QString>` with frameworks. Can you search around the Qt installations for a directory called `QtCore`? If we can add the parent directory of that as an include path, hopefully all the includes will work. Also, can you post the output of `qmake -query` with the framework installation of Qt?

jacquetc replied:
Sorry, I'm holidaying. 

```
cyril@MacBook-Pro-de-Cyril cxx-qt % /Users/cyril/Qt/6.5.2/macos/bin/qmake -v
QMake version 3.1
Using Qt version 6.5.2 in /Users/cyril/Qt/6.5.2/macos/lib
cyril@MacBook-Pro-de-Cyril cxx-qt % /Users/cyril/Qt/6.5.2/macos/bin/qmake -query
QT_SYSROOT:
QT_INSTALL_PREFIX:/Users/cyril/Qt/6.5.2/macos
QT_INSTALL_ARCHDATA:/Users/cyril/Qt/6.5.2/macos
QT_INSTALL_DATA:/Users/cyril/Qt/6.5.2/macos
QT_INSTALL_DOCS:/Users/cyril/Qt/Docs/Qt-6.5.2
QT_INSTALL_HEADERS:/Users/cyril/Qt/6.5.2/macos/include
QT_INSTALL_LIBS:/Users/cyril/Qt/6.5.2/macos/lib
QT_INSTALL_LIBEXECS:/Users/cyril/Qt/6.5.2/macos/libexec
QT_INSTALL_BINS:/Users/cyril/Qt/6.5.2/macos/bin
QT_INSTALL_TESTS:/Users/cyril/Qt/6.5.2/macos/tests
QT_INSTALL_PLUGINS:/Users/cyril/Qt/6.5.2/macos/plugins
QT_INSTALL_QML:/Users/cyril/Qt/6.5.2/macos/qml
QT_INSTALL_TRANSLATIONS:/Users/cyril/Qt/6.5.2/macos/translations
QT_INSTALL_CONFIGURATION:
QT_INSTALL_EXAMPLES:/Users/cyril/Qt/Examples/Qt-6.5.2
QT_INSTALL_DEMOS:/Users/cyril/Qt/Examples/Qt-6.5.2
QT_HOST_PREFIX:/Users/cyril/Qt/6.5.2/macos
QT_HOST_DATA:/Users/cyril/Qt/6.5.2/macos
QT_HOST_BINS:/Users/cyril/Qt/6.5.2/macos/bin
QT_HOST_LIBEXECS:/Users/cyril/Qt/6.5.2/macos/libexec
QT_HOST_LIBS:/Users/cyril/Qt/6.5.2/macos/lib
QMAKE_SPEC:macx-clang
QMAKE_XSPEC:macx-clang
QMAKE_VERSION:3.1
QT_VERSION:6.5.2
```

jacquetc replied:
[qt_files.txt](https://github.com/KDAB/cxx-qt/files/12284284/qt_files.txt)

Be-ing replied:
I looked at the CI output from another projecting using the Qt official binary installer and the include paths are:
```
-isystem /Users/build/Qt/online/6.3.2/macos/lib/QtCore.framework/Headers
-isystem /Users/build/Qt/online/6.3.2/macos/lib/QtCore.framework/Headers/6.3.2
-isystem /Users/build/Qt/online/6.3.2/macos/lib/QtCore.framework/Headers/6.3.2/QtCore
```

So it seems we need:
```
{lib_root_path}/Qt{qt_module}.framework/Headers
{lib_root_path}/Qt{qt_module}.framework/Headers/{version}
{lib_root_path}/Qt{qt_module}.framework/Headers/{version}/{qt_module}
```

jacquetc replied:
Hello,
I'm back, I'll be able to test that this week.

Be-ing replied:
@jacquetc Have you had a chance to test the code I [posted above](https://github.com/KDAB/cxx-qt/pull/630#discussion_r1286329630)?

jacquetc replied:
No yet, I'll take time tomorrow at noon

jacquetc replied:
Tested, exactly the same error.

Be-ing commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -369,10 +399,12 @@ impl QtBuild {
             lib_path, prefix, version_major, qt_module
-        )
+        );
+        println!("cargo:warning=falling back to {}", prl_file);
```
```suggestion
        println!("cargo:warning=Qt .prl file not found at expected path, falling back to {}", prl_file);
```

ahayzen-kdab commented on crates/cxx-qt-lib/build.rs:
```diff
@@ -19,7 +19,8 @@ fn main() {
         qt_modules.push("Qml".to_owned());
     }
 
```
Note that we don't want merge commits in `main` so please prefer using rebase rather than merging from `main`.

jacquetc replied:
Sorry, I realize now that it was not my call to merge/rebase.

ahayzen-kdab replied:
Thanks! Also note if you aren't used to the rebase model you can then rearrange / squash commits together etc (just watch our for merge conflicts :-) ). So one could keep amending the same commit rather than adding new ones for each feedback comment etc, so then each commit can have individual meaning and be atomic in it's purpose.

jacquetc replied:
Thank for your explanations, I'll do my best to commit this way :)

ahayzen-kdab commented on crates/qt-build-utils/src/lib.rs:
```diff
@@ -369,10 +399,12 @@ impl QtBuild {
             lib_path, prefix, version_major, qt_module
-        )
+        );
+        println!("cargo:warning=Qt .prl file not found at expected path, falling back to {}", prl_file);
```
`cargo fmt` fails on this part

jacquetc replied:
formatted

Be-ing commented:
Please check that this branch also builds with Qt installed from Homebrew, which is also shipped as frameworks. You can test that by setting the QMAKE environment variable, for example `QMAKE=/path/to/qmake/from/homebrew cargo run -p qml-minimal-no-cmake`

jacquetc commented:
> QMAKE=/path/to/qmake/from/homebrew cargo run -p qml-minimal-no-cmake

Tested successfully with homebrew's Qt

ahayzen-kdab commented:
Parts of this done in #970 which then passes CI with Apple frameworks from the official Qt installer in #1098, therefore I'm going to close this pull request.
